File: app/src/main/java/com/king/mvvmframe/app/dictionary/DictionaryViewModel.java
package com.king.mvvmframe.app.dictionary;

import android.app.Application;
import android.text.TextUtils;

import com.king.frame.mvvmframe.base.BaseViewModel;
import com.king.frame.mvvmframe.bean.Resource;
import com.king.mvvmframe.R;
import com.king.mvvmframe.bean.DictionaryInfo;
import com.king.mvvmframe.bean.SearchHistory;

import java.util.List;

import androidx.annotation.NonNull;
import androidx.hilt.lifecycle.ViewModelInject;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MediatorLiveData;

/**
 * @author <a href="mailto:jenly1314@gmail.com">Jenly</a>
 */
public class DictionaryViewModel extends BaseViewModel<DictionaryModel> {

    private MediatorLiveData<DictionaryInfo> dictionaryLiveData = new MediatorLiveData<>();

    private MediatorLiveData<List<SearchHistory>> searchHistoryLiveData = new MediatorLiveData<>();

    private LiveData<Resource<DictionaryInfo>> dictionarySource;

    private LiveData<List<SearchHistory>> searchHistorySource;

    @ViewModelInject
    public DictionaryViewModel(@NonNull Application application, DictionaryModel model) {
        super(application, model);
    }

    @Override
    public void onCreate() {
        super.onCreate();
        getSearchHistory(10);
    }

    /**
     * 获取字典信息
     * @param word 要查询的汉子
     */
    public void getDictionaryInfo(String word){
        if(dictionarySource != null){
            dictionaryLiveData.removeSource(dictionarySource);
        }
        dictionarySource = getModel().getDictionaryInfo(word);
        dictionaryLiveData.addSource(dictionarySource, resource -> {
            updateStatus(resource.status);
            if(resource.isSuccess()){//成功
                dictionaryLiveData.postValue(resource.data);
            }else if(resource.isFailure()){//失败
                if(!TextUtils.isEmpty(resource.message)){
                    postMessage(resource.message);
                }else{
                    postMessage(R.string.result_failure);
                }
            }else if(resource.isError()){
                postMessage(resource.error.getMessage());
            }
        });
    }

    /**
     * 获取搜索历史
     * @param count
     */
    public void getSearchHistory(int count){
        if(searchHistorySource!=null){
            searchHistoryLiveData.removeSource(searchHistorySource);
        }
        searchHistorySource = getModel().getSearchHistory(count);
        searchHistoryLiveData.addSource(searchHistorySource,searchHistoryList ->
                searchHistoryLiveData.setValue(searchHistoryList)
        );
    }

    /**
     * 清空历史
     */
    public void deleteAllHistory(){
        getModel().deleteAllHistory();
    }

    public LiveData<DictionaryInfo> getDictionaryLiveData(){
        return dictionaryLiveData;
    }

    public LiveData<List<SearchHistory>> getSearchHistoryLiveData(){
        return searchHistoryLiveData;
    }
}

File: app/src/main/kotlin/com/king/mvvmframe/app/dictionary/DictionaryViewModel.kt
package com.king.mvvmframe.app.dictionary

import android.app.Application
import com.king.mvvmframe.R
import com.king.mvvmframe.app.base.BaseViewModel
import com.king.mvvmframe.data.model.DictionaryInfo
import com.king.mvvmframe.data.repository.DictionaryRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import javax.inject.Inject
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asSharedFlow

/**
 * 示例
 *
 * @author <a href="mailto:jenly1314@gmail.com">Jenly</a>
 * <p>
 * <a href="https://github.com/jenly1314">Follow me</a>
 */
@HiltViewModel
class DictionaryViewModel @Inject constructor(
    private val dictionaryRepository: DictionaryRepository,
    application: Application
) : BaseViewModel(application) {

    private val _dictionaryFlow = MutableStateFlow<DictionaryInfo?>(null)
    val dictionaryFlow = _dictionaryFlow.asSharedFlow()

    /**
     * 获取字典信息
     */
    fun getDictionaryInfo(word: String) {
        launch {
            val result = dictionaryRepository.getDictionaryInfo(word)
            if (isSuccess(result)) {
                _dictionaryFlow.emit(result.data)
            }
        }
    }

    /**
     * 获取搜索历史
     */
    fun getSearchHistoryFlow() = dictionaryRepository.getSearchHistoryFlow()

    /**
     * 删除搜索历史
     */
    fun clearSearchHistory() {
        launch {
            dictionaryRepository.clearSearchHistory()
            sendMessage(R.string.cleared)
        }
    }

}
